conjure-runtime/src/client_cache.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    sync::{Arc, Weak},
+    time::Duration,
+};
+
+use conjure_error::Error;
+use conjure_runtime_config::{ProxyConfig, SecurityConfig};
+use linked_hash_map::LinkedHashMap;
+use parking_lot::Mutex;
+use url::Url;
+
+use crate::{
+    raw::DefaultRawClient, Builder, ClientQos, ClientState, Idempotency, NodeSelectionStrategy,
+    ServerQos, ServiceError, UserAgent,
+};
+
+const MAX_CACHED_CHANNELS: usize = 1_000;
+
+struct CachedState {
+    state: Weak<ClientState<DefaultRawClient>>,
+    id: usize,
+}
+
+struct Inner {
+    cache: LinkedHashMap<Arc<CacheKey>, CachedState>,
+    next_id: usize,
+}
+
+#[derive(Clone)]
+pub struct ClientCache {
```

Comment:
It's probably worth adding some tests for the ClientCache logic, especially since the evictor logic is a bit tricky

Reply:
Yeah. I think the best route is probably to make the cache generic so I can use a different type than DefaultRawClient for testing?

Reply:
Yeah, that sounds right. If you make the cache key generic it might also make it easier to move that closer to the builder logic, where the builder comes with a method to generate the properties of the client